domain/application/service/migration.go

```diff
@@ -295,19 +296,45 @@ func (s *MigrationService) ImportApplication(ctx context.Context, name string, a
 		return errors.Annotatef(err, "creating application args")
 	}
 
-	units := args.Units
-	numUnits := len(units)
-	appArg.Scale = numUnits
+	appArg.Scale = len(args.Units)
+	unitArgs, err := makeUnitArgs(args.Units)
+	if err != nil {
+		return errors.Annotatef(err, "creating unit args")
+	}
+
+	appID, err := s.st.CreateApplication(ctx, name, appArg, nil)
+	if err != nil {
+		return errors.Annotatef(err, "creating application %q", name)
+	}
+	if modelType == model.IAAS {
+		err = s.st.InsertMigratingIAASUnits(ctx, appID, unitArgs...)
+	} else {
+		err = s.st.InsertMigratingCAASUnits(ctx, appID, unitArgs...)
+	}
+	if err != nil {
+		return errors.Annotatef(err, "inserting units for application %q", name)
+	}
+	// TODO(nvinuesa): We should use the service method to set the
+	// constraints because we need validation. For the moment it's not
+	// possible because we need a provider during migration.
+	err = s.st.SetApplicationConstraints(ctx, appID, constraints.DecodeConstraints(args.ApplicationConstraints))
+	if err != nil {
+		return errors.Annotatef(err, "setting application constraints for application %q", name)
+	}
```

Comment:
This can then be lifted into the new ImportApplication and can be done as one transaction.